### PR TITLE
Add documentation for how to register new StorageDrivers

### PR DESCRIPTION
**Pull request by arschles**
This commit adds context-specific documentation on StorageDriver,
StorageDriverFactory, and the factory’s Register func, explaining how
the internal registration mechanism should be used.

This documentation follows from the thread starting at
https://github.com/deis/builder/pull/262/files#r56720200.

cc/ @stevvooe

### CODE REVIEW

**Comment by arschles** [comment]
@GordonTheTurtle done, see below for proof:

```console
ENG000656:Desktop aaronschlesinger$ git clone -b "doc-storage" git@github.com:arschles/docker-distribution.git somewhere
Cloning into 'somewhere'...
remote: Counting objects: 13107, done.
remote: Compressing objects: 100% (25/25), done.
remote: Total 13107 (delta 6), reused 0 (delta 0), pack-reused 13081
Receiving objects: 100% (13107/13107), 8.88 MiB | 749.00 KiB/s, done.
Resolving deltas: 100% (6938/6938), done.
Checking connectivity... done.
ENG000656:Desktop aaronschlesinger$ cd somewhere
ENG000656:somewhere aaronschlesinger$ git commit --amend -s --no-edit
[doc-storage 28757e6] Add documentation for how to register new StorageDrivers
 Date: Fri Mar 18 16:28:42 2016 -0700
 2 files changed, 18 insertions(+), 2 deletions(-)
ENG000656:somewhere aaronschlesinger$ git push -f
warning: push.default is unset; its implicit value has changed in
Git 2.0 from 'matching' to 'simple'. To squelch this message
and maintain the traditional behavior, use:

  git config --global push.default matching

To squelch this message and adopt the new behavior now, use:

  git config --global push.default simple

When push.default is set to 'matching', git will push local branches
to the remote branches that already exist with the same name.

Since Git 2.0, Git defaults to the more conservative 'simple'
behavior, which only pushes the current branch to the corresponding
remote branch that 'git pull' uses to update the current branch.

See 'git help config' and search for 'push.default' for further information.
(the 'simple' mode was introduced in Git 1.7.11. Use the similar mode
'current' instead of 'simple' if you sometimes use older versions of Git)


Counting objects: 8, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (7/7), done.
Writing objects: 100% (8/8), 1.50 KiB | 0 bytes/s, done.
Total 8 (delta 6), reused 1 (delta 0)
To git@github.com:arschles/docker-distribution.git
 + 9a74b45...28757e6 doc-storage -> doc-storage (forced update)
```

File: registry/storage/driver/storagedriver.go [hunk 1]
```diff
@@ -34,7 +34,14 @@ func (version Version) Minor() uint {
 const CurrentVersion Version = "0.1"
 
 // StorageDriver defines methods that a Storage Driver must implement for a
-// filesystem-like key/value object storage.
+// filesystem-like key/value object storage. Storage Drivers are automatically
+// registered via an internal registration mechanism, and generally created
+// via the StorageDriverFactory interface (https://godoc.org/github.com/docker/distribution/registry/storage/driver/factory).
+// See below for an example of how to get a StorageDriver for S3:
+//
+//	import _ "github.com/docker/distribution/registry/storage/driver/s3-aws"
```
Comment (stevvooe) [inline registry/storage/driver/storagedriver.go]: I don't think this example belongs at the top-level package. The use of factories are not required to use storage drivers.

Maybe just a pointer here to the factory package as a hint. This way, we avoid having to maintain this example in two places.

**Comment by stevvooe** [comment]
@arschles Thanks for the PR! A few nits and we can get this in.

**Comment by arschles** [comment]
@stevvooe Thanks for reviewing! I think I addressed the nit at https://github.com/docker/distribution/pull/1546#discussion_r56894811 in 7cf5fdc. Were there any other comments that I missed?

**Comment by stevvooe** [comment]
@arschles Make sure to sign _both_ commits and I think we're good.

**Comment by arschles** [comment]
@stevvooe just amended to sign both commits. thanks!

**Comment by stevvooe** [comment]
LGTM

@RichardScothern @dmcgowan @dmp42

**Comment by dmcgowan** [comment]
LGTM

**Pull request merged**